File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.py
n = int(input())
num = list(map(int, input().split()))
ord = list(map(int, input().split()))
answer = []

def cal(x, y, id):
    if id == 0:
        return x + y
    elif id == 1:
        return x - y
    elif id == 2:
        return x * y
    else:
        if x < 0:
            return -(abs(x)//y)
        else:
            return x//y

def dfs(value, idx, op):
    if idx == n:
        answer.append(value)
        return
    for i, o in enumerate(ord):
        if o:
            ord[i] -= 1
            new_value = cal(value, num[idx],i)
            dfs(new_value, idx + 1, ord)
            ord[i] += 1
dfs(num[0], 1, ord)
print(max(answer))
print(min(answer))

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.java
import java.io.*;
import java.util.*;

public class Main {
    static int N;
    static int[] numbers;       // 입력 숫자
    static int[] operators;     // 연산자 개수: [+,-,*,/]
    static int max = Integer.MIN_VALUE;
    static int min = Integer.MAX_VALUE;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        N = Integer.parseInt(br.readLine());
        numbers = new int[N];
        operators = new int[4];

        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            numbers[i] = Integer.parseInt(st.nextToken());
        }

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < 4; i++) {
            operators[i] = Integer.parseInt(st.nextToken());
        }

        dfs(1, numbers[0]);

        System.out.println(max);
        System.out.println(min);
    }

    static void dfs(int idx, int result) {
        if (idx == N) {
            max = Math.max(max, result);
            min = Math.min(min, result);
            return;
        }

        for (int i = 0; i < 4; i++) {
            if (operators[i] > 0) {
                operators[i]--;

                int next = result;
                if (i == 0) next += numbers[idx];           // 덧셈
                else if (i == 1) next -= numbers[idx];      // 뺄셈
                else if (i == 2) next *= numbers[idx];      // 곱셈
                else if (i == 3) next /= numbers[idx];      // 나눗셈 (정수 나눗셈)

                dfs(idx + 1, next);
                operators[i]++;
            }
        }
    }
}
